src/topology/homotopy/homotopy_group.lean

```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
 -/
 
 import algebraic_topology.fundamental_groupoid.fundamental_group
+import group_theory.eckmann_hilton
+import logic.equiv.transfer_instance
+import algebra.group.ext
 
 /-!
 # `n`th homotopy group
 
-We define the `n`th homotopy group at `x`, `π n x`, as the equivalence classes
+We define the `n`th homotopy group at `x`, `π_n x`, as the equivalence classes
 of functions from the nth dimensional cube to the topological space `X`
 that send the boundary to the base point `x`, up to homotopic equivalence.
 Note that such functions are generalized loops `gen_loop n x`, in particular
 `gen_loop 1 x ≃ path x x`
 
-We show that `π 0 x` is equivalent to the path-conected components, and
-that `π 1 x` is equivalent to the fundamental group at `x`.
+We show that `π_0 x` is equivalent to the path-conected components, and
+that `π_1 x` is equivalent to the fundamental group at `x`.
+We give a group instance using path composition and show commutativity when `n > 1`.
 
 ## definitions
 
-* `gen_loop n x` is the type of continous fuctions `I^n → X` that send the boundary to `x`
-* `homotopy_group n x` denoted `π n x` is the quotient of `gen_loop n x` by homotopy relative
-  to the boundary
+* `gen_loop n x` is the type of continous fuctions `I^n → X` that send the boundary to `x`,
+* `homotopy_group n x` denoted `π_ n x` is the quotient of `gen_loop n x` by homotopy relative
+  to the boundary,
+* group instance `group (π_(n+1) x)`,
+* commutative group instance `comm_group (π_(n+2) x)`.
 
-TODO: show that `π n x` is a group when `n > 0` and abelian when `n > 1`. Show that
-`pi1_equiv_fundamental_group` is an isomorphism of groups.
+TODO: Path-induced homomorphisms. Show that `pi1_equiv_fundamental_group` is a group isomorphism.
+  Examples with 𝕊^n (π_n (𝕊^n) = ℤ, π_m (𝕊^n) trivial for m < n).
+  Actions of π_1 on π_n.
+  Group (up to homotopy) of Ω.
+  Lie algebra: ⁅π_(n+1), π_(m+1)⁆ contained in π_(n+m+1).
 
 -/
 
 open_locale unit_interval topology
+open homeomorph
 
 noncomputable theory
 
 universes u
 variables {X : Type u} [topological_space X]
-variables {n : ℕ} {x : X}
+variables {N : Type*} {x : X}
 
-/--
-The `n`-dimensional cube.
--/
+/-- We allow an arbitrary indexing type `N`, not just `fin n`, in the definition of `cube`, so as to
+  be able to identify facets of a cube to cubes indexed by subtypes of `N`, which is more
+  convenient than using `fin (n-1)`. -/
 @[derive [has_zero, has_one, topological_space]]
-def cube (n : ℕ) : Type := fin n → I
-local notation `I^` := cube
+def cube (N : Type*) := N → I
+local notation `I^` n := cube (fin n)
```

Comment:
Sorry, I know you haven't introduced that, but why have a type synonym for pi types?

Reply:
Essentially, just for legibility, like `N×M` matrices over `X` being functions `N → M → X`.

Reply:
Yes, but there are other ways of improving legibility than introducing new definitions. For example, reducible definitions or notation.

Reply:
I've removed the def and generalized the `I^` notation for arbitrary `N`

Reply:
Yeah, making `Ω^N` a local notation seems reasonable; indeed it's common to write `Ω^n` for iterated loop spaces as in https://en.wikipedia.org/wiki/Spectrum_(topology)#Suspension_spectrum, but since we don't want to write `Ω^fin n` it's wise to keep the `Ω^N` notation local, so we can introduce the `Ω^n` notation when needed.